Task 1 Contest 1

d-tiapkin opened the pull request:
Задача "Ручная сортировка". 

Тяпкин Даниил, 173-2.

d-tiapkin committed:
First task

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+#include <cstddef>
+#include <utility>
+
```
Не надо выделять iostream.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+
+#include <iostream>
+
+namespace odl {
```
Что значит odl?

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+
+        Node(Node* next) :
+                next(next),
+                value()
```
Конструктор по умолчанию вызовется и так, вызывать его явно не обязательно.

explicit.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+class List {
+public:
+    List() {
+        start = new odl::Node<T>(finish);
```
Лучше begin/end или head/tail.

finish изменится через две строчки, ты уверен, что хотел передать в конструктор именно это?

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+public:
+    List() {
+        start = new odl::Node<T>(finish);
+        last = &start;
```
Взяв указатель на член класса, проиграешь при копировании.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    odl::Node<T>* start;
+    odl::Node<T>** last;
+    odl::Node<T>* finish;
+    size_t size_val;
```
size_ лучше, чем size_val. К остальным переменным это тоже относится.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+
+        ListIterator operator++(int) {
+            auto copy = ListIterator(*this);
+            cur = cur->next;
```
Лучше сделать ++*this, чтобы избежать копирования кода.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    }
+
+private:
+    void NodeDelete(odl::Node<T>*& cur) {
```
Во-первых, стоит называть функции корректными фразами: NodeDelete.

Во-вторых, почему принимается ссылка на указатель?

В-третьих, как по названию догадаться, что функция удалит не только узел, но и все последующие?

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+};
+
+template <typename T>
+void sort(const odl::ListIterator<T>& begin, const odl::ListIterator<T>& end, size_t n) {
```
Если функция принимает пару итераторов, не надо отдельно ещё и размер передавать.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    auto mid = begin;
+    size_t sz;
+    for (sz = 0; 2*sz < n; ++sz) {
+        mid++;
```
Префиксный инкремент быстрее.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    odl::Node<T>* start;
+    odl::Node<T>** last;
```
В чём вообще смысл last? Кажется, двух указателей должно быть достаточно.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    sort(begin, mid, sz);
+    sort(mid, end, n - sz);
+
+    List<T> s;
```
Комментарий ко всему, что дальше: стоит сделать так, чтобы не копировать значения в узлах, а только перекидывать указатели.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+        std::cout << elem << ' ';
+    }
+    std::cout << '\n';
+}
```
Пустая строка в конце файла.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    List<int> lst;
+    int n;
+    std::cin >> n;
+    for (size_t i = 0; i != n; ++i) {
```
Сравнение знаковых и беззнаковых.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+    for (; it1 != end; ++it1, ++it2) {
+        *it1 = *it2;
+    }
+
```
Лишняя пустая строка здесь и после функции.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,169 @@
+
+    it1 = begin;
+    it2 = s.begin();
+    for (; it1 != end; ++it1, ++it2) {
```
Лучше while (it1 != end) *it1++ = *it2++;

d-tiapkin committed:
Some changes

ifsmirnov commented on source:
```diff
@@ -1,48 +1,58 @@
         }
 
-        ListIterator operator++(int) {
+        const ListIterator& operator++(int) {
```
Возвращаешь ссылку на временную переменную, ты уверен?

ifsmirnov commented on source:
```diff
@@ -57,113 +67,122 @@ namespace odl {
 private:
-    void NodeDelete(odl::Node<T>*& cur) {
+    // Delete all nodes from the list.
+    void NodeClear(fwd_list::Node<T>* cur) {
```
DeleteAllNodes, например. Всё ещё непонятно из названия, что должна делать функция.

ifsmirnov commented on source:
```diff
@@ -57,113 +67,122 @@ namespace odl {
 
 template <typename T>
-void sort(const odl::ListIterator<T>& begin, const odl::ListIterator<T>& end, size_t n) {
+void sort(const fwd_list::ListIterator<T>& begin, const fwd_list::ListIterator<T>& end) {
```
Итераторы обычно считаются лёгкими типами, их можно передавать по значению.

ifsmirnov commented on source:
```diff
@@ -57,113 +67,122 @@ namespace odl {
 
     for (; it1 != mid; ++it1) {
-        s.push_back(*it1);
+        cur->next = new fwd_list::Node<T>(nullptr, it1.ptr_value());
```
Этот и следующий блоки кода стоит перенести внутрь предыдущего цикла, добавив в условие проверку на пустоту.

ifsmirnov commented on source:
```diff
@@ -57,113 +67,122 @@ namespace odl {
     auto it1 = begin;
     auto it2 = mid;
 
     for (; it1 != mid && it2 != end;) {
```
while

ifsmirnov commented on source:
```diff
@@ -57,113 +67,122 @@ namespace odl {
     for (; it1 != mid && it2 != end;) {
         if (*it1 > *it2) {
-            s.push_back(*it2);
+            cur->next = new fwd_list::Node<T>(nullptr, it2.ptr_value());
```
И ты всё ещё вызываешь new бесконечно много раз.

d-tiapkin committed:
Rebuild structure

I find that changing pointers under iterators ruins philosophy of iterators in main. 
The sort function was been placed into class body and now it's working twice faster and required more than twice memory less.

ifsmirnov commented on source:
```diff
@@ -0,0 +1,177 @@
+
+    fwd_list::Node<T>* head;
+    fwd_list::Node<T>* tail;
+    size_t sz;
```
head_, tail_, size_. Ну уж никак не sz.

ifsmirnov commented on source:
```diff
@@ -7,23 +7,16 @@ namespace fwd_list {
         explicit Node(Node* next) :
                 next(next),
-                value(nullptr)
+                value()
```
up. value() писать не надо.

ifsmirnov reviewed (approved):
Поговорили голосом.
